# Lab Three: Extending Logistic Regression
## Caleb Moore, Blake Gebhardt, Christian Gould
dataset: https://www.kaggle.com/datasets/open-powerlifting/powerlifting-database

In [103]:
# Imports
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn import datasets

In [104]:
# Notebook setup
from IPython.display import HTML
HTML('''<script>
code_show_err=false; 
function code_toggle_err() {
 if (code_show_err){
 $('div.output_stderr').hide();
 } else {
 $('div.output_stderr').show();
 }
 code_show_err = !code_show_err
} 
$( document ).ready(code_toggle_err);
</script>
To toggle on/off output_stderr, click <a href="javascript:code_toggle_err()">here</a>.''')

# Preparation and Overview (3 points total)
* [2 points] Explain the task and what business-case or use-case it is designed to solve (or designed to investigate). Detail exactly what the classification task is and what parties would be interested in the results. For example, would the model be deployed or used mostly for offline analysis? As in previous labs, also detail how good the classifier needs to perform in order to be useful.

# Background
### Where did our data come from?
* Our data came from the Open Powerlifting Databse. They are trying to create a database of powerlifting competitions for the public domain. The dataset that we are using here is a snapshot of what the database looked like as of April 2019. Powerlifting is a sport in which competitors compete to lift the most weight for their class in three separate barbell lifts: the Squat, Bench, and Deadlift. The dataset that we have here contains information about the competitors, the competitions, and the lifts that they performed.

### What could we do with this data?
* There is a lot of data that we can use here, but what we are more focused on is the prediction of the weight class of each person. If we can use data to predict with a very high accuracy what weight class a person should be in, it could be helpful in determining if someone is cheating or not. People could lie about their weight if there is not a weighing time. This could also show if someone is likely to be taking steroids. If they are classified as being far above the weight class than what they actually are, then investigators could look into the possiblity of them taking steroids.

### What is the task we are trying to solve?
* Given all the information that we have, specifically:
  * Sex
    * Sex has a large impact on how much the lifter can lift
  * Equipment
    * If the lifter uses straps or other apparatus, it can affect how much they can lift
  * Age
    * Age has a large impact on how much the lifter can lift
  * Division
    * A higher skill division can mean that the lifter can lift more
  * BodyweightKg
    * The actual weight of the lifter is a large factor in how much they can lift
  * Best3SquatKg
    * The best squat that the lifter did in 3 attempts
  * Best3BenchKg
    * The best bench that the lifter did in 3 attempts
  * Best3DeadliftKg
    * The best deadlift that the lifter did in 3 attempts
  * Tested
    * Whether or not the lifter was tested for drugs
  * Federation
    * The federation that held the competition

We want to be able to reliably predict what Weight Class that person should be in. The target column for that data is WeightClassKg.

The rest of the columns are not as important to us, so we will be dropping them.

The columns to be dropped are:
* Name
  * The name of the lifter is irrelevant to the weight class that they should be in
* Event
  * The federation that held the event is irrelevant to the weight class that they should be in
* TotalKg
  * The total weight lifted by the lifter. This is redundant since it is just the sum of the best squat, best bench, and best deadlift, which we already have
* AgeClass
  * Because we already have the age of the lifter, this column is redundant
* Squat1Kg - Squat4Kg
  * These are the individual squat attempts, we only really care about their best one
* Bench1Kg - Bench4Kg
  * These are the individual bench attempts, we only really care about their best one
* Deadlift1Kg - Deadlift4Kg
  * These are the individual deadlift attempts, we only really care about their best one
* Place
  * We don't need to know where the competition was held, just the results
* Wilks
  * This is a calculation of the total weight lifted in relation to the weight of the lifter, which is information we are already accounting for, so it would be redundant
* McCulloch
  * This is another calculation method like Wilks, and for the same reason as Wilks, we will not need this.
* Glossbrenner
  * This is another calculation method like Wilks, and for the same reason as Wilks, we will not need this.
* IPFPoints
  * This is a calculation of the total weight lifted in relation to the weight of the lifter, which is information we are already accounting for, so it would be redundant
* MeetCountry
  * We don't need to know where the competition was held, just the results
* MeetState
  * We don't need to know where the competition was held, just the results
* MeetName
  * We don't need to know the name of the meet, just what was lifted
* Country
  * We don't need to know where the lifter is from, just what they lifted
* Date
  * We don't need to know when the competition was held, just what was lifted

### What are we aiming for

* If we want our classifier to be reliable, we need to have at least 90% accuracy. If we can get to that, then we can be confident that we can use this classifier to determine if someone is possibly cheating or taking steroids. That level of accuracy would be very useful for investigators. 

### Things to note
* The full dataset has over 1.4 million rows of data, and that simply is more than we would like to operate on. We are going to shrink our dataset size down to 100k rows and deal with that data instead, in the interest of runtime


In [105]:
cols_to_drop = [
    'Name',
    'Event',
    'TotalKg',
    'AgeClass',
    'Squat1Kg',
    'Squat2Kg',
    'Squat3Kg',
    'Squat4Kg',
    'Bench1Kg',
    'Bench2Kg',
    'Bench3Kg',
    'Bench4Kg',
    'Deadlift1Kg',
    'Deadlift2Kg',
    'Deadlift3Kg',
    'Deadlift4Kg',
    'Place',
    'Wilks',
    'McCulloch',
    'Glossbrenner',
    'IPFPoints',
    'MeetCountry',
    'MeetState',
    'MeetName',
    'Country',
    'Date',
]

[.5 points] (mostly the same processes as from previous labs) Define and prepare your class variables. Use proper variable representations (int, float, one-hot, etc.). Use pre-processing methods (as needed) for dimensionality reduction, scaling, etc. Remove variables that are not needed/useful for the analysis. Describe the final dataset that is used for classification/regression (include a description of any newly formed variables you created). Provide a breakdown of the variables after preprocessing (such as the mean, std, etc. for all variables, including numeric and categorical).

# Initial Data Exploration
* We will start by importing the data and looking at the first few rows to get a feel for what we are working with.

* Some of the rows have NaN values, specifically in the 3 main lifts. We will give those NaN values a value of 0, since those people did not lift anything. Some of the meets did not have competitors do all 3 lifts, and we want to keep those rows, so we will just give them a value of 0 for that specific lift

* We will also drop the rows that have NaN values in the WeightClassKg column, since we want to predict that column.

* We took a look at how many weight classes there were, and there ended up being 13. A couple of the weight classes were in a different format, where it was 140+ or 110+ instead of the typical weight class, and we will leave those unaffected since some federations use that format.

In [106]:
# lets look at the data
df = pd.read_csv('./data/openpowerlifting.csv', nrows=100000)
# Drop the columns we don't need
df = df.drop(cols_to_drop, axis=1)
print(df.shape)
df.head()

(100000, 11)


Sex Equipment   Age Division  BodyweightKg WeightClassKg  Best3SquatKg  \
0   F     Wraps  29.0     F-OR          59.8            60         105.0   
1   F     Wraps  29.0     F-OR          58.5            60         120.0   
2   F       Raw  40.0     F-OR          55.4            56           NaN   
3   F     Wraps  23.0     F-OR          60.0            60         105.0   
4   F     Wraps  45.0     F-OR         104.0           110         140.0   

   Best3BenchKg  Best3DeadliftKg Tested Federation  
0          55.0            130.0    NaN    GPC-AUS  
1          67.5            145.0    NaN    GPC-AUS  
2          32.5              NaN    NaN    GPC-AUS  
3          72.5            132.5    NaN    GPC-AUS  
4          80.0            170.0    NaN    GPC-AUS

### Look at the types of Weight Classes
* There are a ton of weight classes, and we want to reduce the number of classes that we have. We will do this by combining weight classes into categories of 10s. For example, all values 111-120 will become 120, and all values 121-130 will become 130. This will reduce the number of classes that we have, and it will also make it easier to predict the weight class.
* Therefore we ended with a total of 12 possible weight classes, and we will use those as our target column.
* We decided to remove any rows that have a + at the end, since if the specification is something like 110+, then it is not a specific weight class, and we want to be able to predict a specific weight class.

In [107]:
# Lets look at some unique values that might be worth encoding, specifically our weight classes
print(df['WeightClassKg'].unique())
print(len(df['WeightClassKg'].unique()))

# Remove all rows with a NaN value for weight class, since that is the value we are trying to predict
df = df.dropna(subset=['WeightClassKg'])

# All the rows with a + at the end do not give us enough information to predict the weight class.
# We will remove them
df = df[~df['WeightClassKg'].str.contains('\+')]

# Convert to float
df['WeightClassKg'] = df['WeightClassKg'].astype(float)

# Convert all values between factors of ten rounded up to the nearest factor of ten
df['WeightClassKg'] = df['WeightClassKg'].apply(lambda x: 10 * round(x / 10))

# Make it a string again so that it can be treated as a categorical variable
df['WeightClassKg'] = df['WeightClassKg'].astype(str)

# look at it real quick
print(df['WeightClassKg'].unique())

['60' '56' '110' '75' '82.5' '52' '67.5' '90' '110+' '125' '100' '140'
 '140+' '48' '90+' '44' nan '105' '74' '93' '120+' '120' '63' '83' '84'
 '57' '72' '84+' '47' '59' '66' '53' '125+' '43' '100+' '90.7' '90.7+'
 '36' '40' '46' '49' '75+' '82.5+' '52+' '145' '145+' '72+' '93+' '60+'
 '50' '65' '80' '80+' '67.5+' '105+' '63+' '35' '155' '39' '155+' '68'
 '50.5' '55.5' '58.5' '70' '47.5' '136' '136+' '55' '54.4' '58.9' '63.5'
 '68.9' '78.9' '78.9+' '56.7' '83.9' '102' '113.4' '127' '127+' '117.5'
 '101' '103' '67' '82' '36.2' '52.5' '72.5' '81.6' '104.3' '117.9'
 '117.9+' '79.3' '113.4+' '49.9' '57.1' '70.3' '77.1' '92.9' '68.5' '69.5'
 '70.5' '71.5' '73.5' '74.5' '75.5' '76.5' '77.5' '78.5' '79.5' '80.5'
 '81.5']
113
['60' '110' '80' '50' '70' '90' '120' '100' '140' '40' '160' '130']


### Tested Column
* This gives us information about whether the lifter was tested for drugs or not. We will convert this to a boolean column, where 1 means that they were tested, and 0 means that they were not tested. 

In [108]:
print(df['Tested'].unique())
# Convert NaN to 0, and 'Yes' to 1
df['Tested'] = df['Tested'].fillna(0)
df['Tested'] = df['Tested'].replace('Yes', 1)

df.head()

[nan 'Yes']


Sex Equipment   Age Division  BodyweightKg WeightClassKg  Best3SquatKg  \
0   F     Wraps  29.0     F-OR          59.8            60         105.0   
1   F     Wraps  29.0     F-OR          58.5            60         120.0   
2   F       Raw  40.0     F-OR          55.4            60           NaN   
3   F     Wraps  23.0     F-OR          60.0            60         105.0   
4   F     Wraps  45.0     F-OR         104.0           110         140.0   

   Best3BenchKg  Best3DeadliftKg  Tested Federation  
0          55.0            130.0       0    GPC-AUS  
1          67.5            145.0       0    GPC-AUS  
2          32.5              NaN       0    GPC-AUS  
3          72.5            132.5       0    GPC-AUS  
4          80.0            170.0       0    GPC-AUS

### Look at the types of Divisions
  * We can turn these into one hot encoded columns, since they are representative of different categories.

In [109]:
print(df['Division'].unique())

# One hot encode that column
df = pd.get_dummies(df, columns=['Division'])
df.head()

['F-OR' 'M-OR' 'M-OE' 'F-OE' 'F-GR' 'F-JR' 'F-T3R' 'F-T2R' 'F-M5R' 'F-M1R'
 'F-M2R' 'F-T1R' 'F-M3R' 'M-JR' 'M-M3R' 'M-M1R' 'M-M2R' 'M-T2R' 'M-T1R'
 'M-M6R' 'M-T3R' 'M-M4R' 'M-M5R' 'M-TR' 'M-JE' 'F-TR' 'F-M4R' 'F-M7R'
 'M-M8R' 'Youth A' 'Juniors' 'Open' 'Masters 1' 'Masters 4' 'Masters 2'
 'Youth B' 'Masters 3' 'Sub-Juniors' 'Youth' 'Teen' 'M1' 'M5' 'M4' 'M2'
 'M3' nan 'Teen 18-19' 'Juniors 20-23' 'Teen 16-17' 'Masters 40-44'
 'Masters 50-54' 'Teen 3' 'Teen 2' 'Masters' 'Teen 1' 'Masters 60-64'
 'Submasters 35-39' 'Masters 55-59' 'Police/Fire' 'Teen 13-15'
 'Submasters 33-39' 'Masters 45-49' 'Masters 65-69' 'Masters 70-74'
 'Juniors 19-20' 'Military' 'Teen 15-16' 'Juniors 21-23' 'Masters 75-79'
 'Teen 14-15' 'Masters 80+' 'Masters 40+' 'M-O' 'F-O' 'F-J' 'F-T2' 'F-T1'
 'F-M1' 'M-T1' 'F-M2' 'M-T2' 'M-J' 'M-M1' 'M-T3' 'M-M5' 'M-M2' 'M-EM6'
 'M-M4' 'M-M3' 'M-SM' 'M-M6' 'F-SM' 'M-EM2' 'M-EM1' 'M-H' 'M-M' 'M-B'
 'F-H' 'F-L' 'M-L' 'F-MW' 'F-LW' 'M-LW' 'M-MW' 'M-HW' 'F-HW' 'M-Hw' 'F-B'
 'M-HWB'

Sex Equipment   Age  BodyweightKg WeightClassKg  Best3SquatKg  Best3BenchKg  \
0   F     Wraps  29.0          59.8            60         105.0          55.0   
1   F     Wraps  29.0          58.5            60         120.0          67.5   
2   F       Raw  40.0          55.4            60           NaN          32.5   
3   F     Wraps  23.0          60.0            60         105.0          72.5   
4   F     Wraps  45.0         104.0           110         140.0          80.0   

   Best3DeadliftKg  Tested Federation  ...  Division_Youth A  \
0            130.0       0    GPC-AUS  ...                 0   
1            145.0       0    GPC-AUS  ...                 0   
2              NaN       0    GPC-AUS  ...                 0   
3            132.5       0    GPC-AUS  ...                 0   
4            170.0       0    GPC-AUS  ...                 0   

   Division_Youth B  Division_law  Division_pro Masters 55-59  \
0                 0             0                           0   
1                 0             0                           0   
2                 0             0                           0   
3                 0             0                           0   
4                 0             0                           0   

   Division_Девушки  Division_М1  Division_М2  Division_М3  Division_Юниорки  \
0                 0            0            0            0                 0   
1                 0            0            0            0                 0   
2                 0            0            0            0                 0   
3                 0            0            0            0                 0   
4                 0            0            0            0                 0   

   Division_Юноши  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 436 columns]

### Federation
* Federations clearly are a categorical datatype, so we will one hot encode them as well.

In [110]:
print(df['Federation'].unique())

# One Hot Encode the Federation column
df = pd.get_dummies(df, columns=['Federation'])
df.head()

['GPC-AUS' 'BVDK' 'GPU' 'GPA-CRO' 'USPF' 'RhinoPC' 'RUPC' 'ACHIPO'
 'USSports' 'UkrainePF' 'UkrainePA' 'WPF' 'USA-UA' 'ADFPA' 'JPA'
 'WPC-Germany' 'PLZS' 'NORCAL' 'WelshPA' 'SSA' 'APA' 'WPA' 'WRPF-KAZ'
 'WUAP-AUT' 'WRPF-POL' 'IPL-NZ' 'APF' 'USMilAbroad' 'AAU' 'BAWLA' 'IPF'
 'PA' 'NZPF' 'AusPF' 'WDFPF' 'IPA' 'WNPF' 'SCI' 'WPC' 'AAPLF' 'SLP'
 'IrishPF' 'CAPO' 'UPA' 'IBSA' 'WABDL' 'ScottishPL' 'EPF' 'AIWBPA' 'BDFPA'
 'NASA' 'BPO' 'USAPL' 'OceaniaPF' 'FPR' 'USARawBP' 'AusDFPF' 'THSPA'
 'BBDD' 'BPF' 'WBC']


Sex Equipment   Age  BodyweightKg WeightClassKg  Best3SquatKg  Best3BenchKg  \
0   F     Wraps  29.0          59.8            60         105.0          55.0   
1   F     Wraps  29.0          58.5            60         120.0          67.5   
2   F       Raw  40.0          55.4            60           NaN          32.5   
3   F     Wraps  23.0          60.0            60         105.0          72.5   
4   F     Wraps  45.0         104.0           110         140.0          80.0   

   Best3DeadliftKg  Tested  Division_0/60  ...  Federation_WDFPF  \
0            130.0       0              0  ...                 0   
1            145.0       0              0  ...                 0   
2              NaN       0              0  ...                 0   
3            132.5       0              0  ...                 0   
4            170.0       0              0  ...                 0   

   Federation_WNPF  Federation_WPA  Federation_WPC  Federation_WPC-Germany  \
0                0               0               0                       0   
1                0               0               0                       0   
2                0               0               0                       0   
3                0               0               0                       0   
4                0               0               0                       0   

   Federation_WPF  Federation_WRPF-KAZ  Federation_WRPF-POL  \
0               0                    0                    0   
1               0                    0                    0   
2               0                    0                    0   
3               0                    0                    0   
4               0                    0                    0   

   Federation_WUAP-AUT  Federation_WelshPA  
0                    0                   0  
1                    0                   0  
2                    0                   0  
3                    0                   0  
4                    0                   0  

[5 rows x 496 columns]

### Look at the Equipment
* Since there are only a few options, we can one-hot encode that column

In [111]:
print(df['Equipment'].unique())

# One hot encode that column
df = pd.get_dummies(df, columns=['Equipment'])
df.head()

['Wraps' 'Raw' 'Single-ply' 'Multi-ply' 'Straps']


Sex   Age  BodyweightKg WeightClassKg  Best3SquatKg  Best3BenchKg  \
0   F  29.0          59.8            60         105.0          55.0   
1   F  29.0          58.5            60         120.0          67.5   
2   F  40.0          55.4            60           NaN          32.5   
3   F  23.0          60.0            60         105.0          72.5   
4   F  45.0         104.0           110         140.0          80.0   

   Best3DeadliftKg  Tested  Division_0/60  Division_13-16  ...  \
0            130.0       0              0               0  ...   
1            145.0       0              0               0  ...   
2              NaN       0              0               0  ...   
3            132.5       0              0               0  ...   
4            170.0       0              0               0  ...   

   Federation_WPF  Federation_WRPF-KAZ  Federation_WRPF-POL  \
0               0                    0                    0   
1               0                    0                    0   
2               0                    0                    0   
3               0                    0                    0   
4               0                    0                    0   

   Federation_WUAP-AUT  Federation_WelshPA  Equipment_Multi-ply  \
0                    0                   0                    0   
1                    0                   0                    0   
2                    0                   0                    0   
3                    0                   0                    0   
4                    0                   0                    0   

   Equipment_Raw  Equipment_Single-ply  Equipment_Straps  Equipment_Wraps  
0              0                     0                 0                1  
1              0                     0                 0                1  
2              1                     0                 0                0  
3              0                     0                 0                1  
4              0                     0                 0                1  

[5 rows x 500 columns]

### Sex
* Since there are only two options here, we will one-hot encode this column as well

In [112]:
print(df['Sex'].unique())

# One hot encode that column
df = pd.get_dummies(df, columns=['Sex'])
df.head()

['F' 'M']


Age  BodyweightKg WeightClassKg  Best3SquatKg  Best3BenchKg  \
0  29.0          59.8            60         105.0          55.0   
1  29.0          58.5            60         120.0          67.5   
2  40.0          55.4            60           NaN          32.5   
3  23.0          60.0            60         105.0          72.5   
4  45.0         104.0           110         140.0          80.0   

   Best3DeadliftKg  Tested  Division_0/60  Division_13-16  Division_14-16  \
0            130.0       0              0               0               0   
1            145.0       0              0               0               0   
2              NaN       0              0               0               0   
3            132.5       0              0               0               0   
4            170.0       0              0               0               0   

   ...  Federation_WRPF-POL  Federation_WUAP-AUT  Federation_WelshPA  \
0  ...                    0                    0                   0   
1  ...                    0                    0                   0   
2  ...                    0                    0                   0   
3  ...                    0                    0                   0   
4  ...                    0                    0                   0   

   Equipment_Multi-ply  Equipment_Raw  Equipment_Single-ply  Equipment_Straps  \
0                    0              0                     0                 0   
1                    0              0                     0                 0   
2                    0              1                     0                 0   
3                    0              0                     0                 0   
4                    0              0                     0                 0   

   Equipment_Wraps  Sex_F  Sex_M  
0                1      1      0  
1                1      1      0  
2                0      1      0  
3                1      1      0  
4                1      1      0  

[5 rows x 501 columns]

# Removing NaN values in the 3 main lifts
* Some of the competitions made it so that the competitors did not have to compete in all 3 lifts, so we will give those NaN values a value of 0, since those people did not lift anything.

In [113]:
# Set the NaN values for the lifts to 0
df['Best3SquatKg'] = df['Best3SquatKg'].fillna(0)
df['Best3BenchKg'] = df['Best3BenchKg'].fillna(0)
df['Best3DeadliftKg'] = df['Best3DeadliftKg'].fillna(0)

### Scaling the data
* We will scale the data using the StandardScaler, since it is a good way to scale the data without distorting the shape of the data.

* The types of data that we need to scale are the Age, BodyweightKg, Best3SquatKg, Best3BenchKg, and Best3DeadliftKg columns. We will scale those columns, and then we will concatenate them with the other columns that we have already one-hot encoded.

In [114]:
# Declare the StandardScaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

# Scale the Age Column
df['Age'] = scaler.fit_transform(df['Age'].values.reshape(-1, 1))

# Scale the BodyweightKg Column
df['BodyweightKg'] = scaler.fit_transform(df['BodyweightKg'].values.reshape(-1, 1))

# Scale the WeightClassKg Column
df['WeightClassKg'] = scaler.fit_transform(df['WeightClassKg'].values.reshape(-1, 1))

# Scale the Best3SquatKg Column
df['Best3SquatKg'] = scaler.fit_transform(df['Best3SquatKg'].values.reshape(-1, 1))

# Scale the Best3BenchKg Column
df['Best3BenchKg'] = scaler.fit_transform(df['Best3BenchKg'].values.reshape(-1, 1))

# Scale the Best3DeadliftKg Column
df['Best3DeadliftKg'] = scaler.fit_transform(df['Best3DeadliftKg'].values.reshape(-1, 1))

df.head()


Age  BodyweightKg  WeightClassKg  Best3SquatKg  Best3BenchKg  \
0 -0.159545     -1.037196      -1.114332     -0.210709     -0.969921   
1 -0.159545     -1.100642      -1.114332     -0.072317     -0.769722   
2  0.635782     -1.251937      -1.114332     -1.179451     -1.330280   
3 -0.593359     -1.027435      -1.114332     -0.210709     -0.689642   
4  0.997294      1.119972       1.348988      0.112206     -0.569523   

   Best3DeadliftKg  Tested  Division_0/60  Division_13-16  Division_14-16  \
0        -0.140969       0              0               0               0   
1        -0.000452       0              0               0               0   
2        -1.358784       0              0               0               0   
3        -0.117550       0              0               0               0   
4         0.233743       0              0               0               0   

   ...  Federation_WRPF-POL  Federation_WUAP-AUT  Federation_WelshPA  \
0  ...                    0                    0                   0   
1  ...                    0                    0                   0   
2  ...                    0                    0                   0   
3  ...                    0                    0                   0   
4  ...                    0                    0                   0   

   Equipment_Multi-ply  Equipment_Raw  Equipment_Single-ply  Equipment_Straps  \
0                    0              0                     0                 0   
1                    0              0                     0                 0   
2                    0              1                     0                 0   
3                    0              0                     0                 0   
4                    0              0                     0                 0   

   Equipment_Wraps  Sex_F  Sex_M  
0                1      1      0  
1                1      1      0  
2                0      1      0  
3                1      1      0  
4                1      1      0  

[5 rows x 501 columns]

[.5 points] Divide your data into training and testing data using an 80% training and 20% testing split. Use the cross validation modules that are part of scikit-learn. Argue "for" or "against" splitting your data using an 80/20 split. That is, why is the 80/20 split appropriate (or not) for your dataset?

* Using an 80/20 split makes sense, as we have a large dataset with a lot of features. We want to make sure that we have enough data to train our model, but we also want to make sure that we have enough data to test it as well. We don't want to overfit our model as a result of using too little data, so we will stick with the 80/20 split.

In [115]:
print('original shapes')
print('X:', df.shape)
print('y:', train_stays.shape)
print()

X_train, X_test, y_train, y_test = train_test_split(df, train_stays, test_size=0.2, random_state=0)

print('train shapes')
print('X:', X_train.shape)
print('y', y_train.shape)
print()

print('test shapes')
print('X:', X_test.shape)
print('y:', y_test.shape)
# clf = svm.SVC(kernel='linear', C=1).fit(X_train, y_train)
# clf.score(X_test, y_test)

original shapes
X: (94413, 501)


NameError: name 'train_stays' is not defined

In [ ]:
X, y = datasets.load_iris(return_X_y=True)
print(X)
y

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

# Modeling (5 points total)
The implementation of logistic regression must be written only from the examples given to you by the instructor. No credit will be assigned to teams that copy implementations from another source, regardless of if the code is properly cited. 

[2 points] Create a custom, one-versus-all logistic regression classifier using numpy and scipy to optimize. Use object oriented conventions identical to scikit-learn. You should start with the template developed by the instructor in the course. You should add the following functionality to the logistic regression classifier:

* Ability to choose optimization technique when class is instantiated: either steepest ascent, stochastic gradient ascent, and {Newton's method/Quasi Newton methods}. 
* Update the gradient calculation to include a customizable regularization term (either using no regularization, L1 regularization, L2 regularization, or both L1 and L2 regularization). Associate a cost with the regularization term, "C", that can be adjusted when the class is instantiated.

In [ ]:
"""
"""
class LRC:
    def __init__(self, learning_rate=0.01, n_iters=1000):
        self.lr = learning_rate
        # Number of iterations is initially 1000
        self.n_iters = n_iters
        # The weights will be made later as well as the bias, whenever we fit
        self.weights = None
        self.bias = None

    """
    Fit function.
    X: The feature matrix
    y: The target vector
    """
    def fit(self, X, y):
        # Number of samples and features
        n_samples, n_features = X.shape

        # set the beginning to have zeros for all the weights
        self.weights = np.zeros(n_features)
        # set the bias to 0 initially
        self.bias = 0

        # gradient descent
        # Used an _ since we don't need to use the iterator, we just need this number of iterations
        for _ in range(self.n_iters):
          # First thing first, calculate the dot product of the weights and the features
            linear_model = np.dot(X, self.weights) + self.bias
          # Then we pass it through the sigmoid function
            y_predicted = self.sigmoid(linear_model)

          # Then we calculate the derivatives
            dw = (1 / n_samples) * np.dot(X.T, (y_predicted - y))
            db = (1 / n_samples) * np.sum(y_predicted - y)

          # Then we update the weights and the bias accordingly
            self.weights -= self.lr * dw
            self.bias -= self.lr * db

    # Function to predict the values
    def predict(self, X):
      # Again, like gradient descent, we calculate the dot product and add the bias
        linear_model = np.dot(X, self.weights) + self.bias
      # Then we pass it through the sigmoid function
        y_predicted = self.sigmoid(linear_model)
      # Then we convert the values to 0 or 1
        y_predicted_cls = [1 if i > 0.5 else 0 for i in y_predicted]
        return y_predicted_cls

    # Basic sigmoid function found online:
    # https://en.wikipedia.org/wiki/Sigmoid_function
    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

[1.5 points] Train your classifier to achieve good generalization performance. That is, adjust the optimization technique and the value of the regularization term(s) "C" to achieve the best performance on your test set. Visualize the performance of the classifier versus the parameters you investigated. Is your method of selecting parameters justified? That is, do you think there is any "data snooping" involved with this method of selecting parameters?

[1.5 points] Compare the performance of your "best" logistic regression optimization procedure to the procedure used in scikit-learn. Visualize the performance differences in terms of training time and classification performance. Discuss the results. 

# Deployment (1 points total)
* Which implementation of logistic regression would you advise be used in a deployed machine learning model, your implementation or scikit-learn (or other third party)? Why?

# Exceptional Work (1 points total)
* You have free reign to provide additional analyses. One idea: Update the code to use either "one-versus-all" or "one-versus-one" extensions of binary to multi-class classification. 